dockers/docker-fpm-frr/bgpd.conf.default.j2

```diff
@@ -36,7 +37,53 @@ route-map ISOLATE permit 10
 route-map set-next-hop-global-v6 permit 10
   set ipv6 next-hop prefer-global
 !
+{% set vrf_lo_intf = {} %}
+{% for (name, prefix) in LOOPBACK_INTERFACE|pfx_filter %}
+{%   set intf_vrf = 'default' %}
+{%   if LOOPBACK_INTERFACE.has_key(name) and LOOPBACK_INTERFACE[name].has_key('vrf_name') %}
+{%     set intf_vrf = LOOPBACK_INTERFACE[name]['vrf_name'] %}
+{%   endif %}
+{%   if vrf_lo_intf.has_key(intf_vrf) or vrf_lo_intf.update({intf_vrf:[]}) %}
+{%   endif %}
+{%   if name not in vrf_lo_intf[intf_vrf] and vrf_lo_intf[intf_vrf].append(name) %}
+{%   endif %}
+{% endfor %}
+{% set vrf_routers = [] %}
+{% set vrf_bgp_nbr = {} %}
+{% for nbr_key in BGP_NEIGHBOR %}
+{%   if nbr_key | length == 2 %}
+{%     set nbr_vrf = nbr_key[0] %}
+{%     set nbr_addr = nbr_key[1] %}
+{%   else %}
+{%     set nbr_vrf = 'default' %}
+{%     set nbr_addr = nbr_key %}
+{%   endif %}
+{%   if nbr_vrf in vrf_routers or vrf_routers.append(nbr_vrf) %}
+{%   endif %}
+{%   if vrf_bgp_nbr.has_key(nbr_vrf) or vrf_bgp_nbr.update({nbr_vrf:{}}) %}
+{%   endif %}
+{%   if vrf_bgp_nbr[nbr_vrf].update({nbr_addr:BGP_NEIGHBOR[nbr_key]}) %}
+{%   endif %}
+{% endfor %}
+{% set vrf_bgp_pr = {} %}
+{% for pr_key in BGP_PEER_RANGE %}
+{%   set pr_vrf = 'default' %}
+{%   if BGP_PEER_RANGE[pr_key].has_key('vrf_name') %}
+{%     set pr_vrf = BGP_PEER_RANGE[pr_key]['vrf_name'] %}
+{%   endif %}
+{%   if pr_vrf not in vrf_routers and vrf_routers.append(pr_vrf) %}
+{%   endif %}
+{%   if vrf_bgp_pr.has_key(pr_vrf) or vrf_bgp_pr.update({pr_vrf:{}}) %}
+{%   endif %}
+{%   if vrf_bgp_pr[pr_vrf].update({pr_key:BGP_PEER_RANGE[pr_key]}) %}
+{%   endif %}
+{% endfor %}
```

Comment:
Is it possible to put all the code inside of bgpcfgd?
Currently this change put a huge amount of logic inside of the jinja2.
Can we put all the logic into bgpcfgd?

Reply:
Not just peer_range, there are configs even peer. If to aware vrf in conf, we have to loop all vrfs. Now vrfs get from BGP_NEIGHBOR and BGP_PEER_RANGE. Theoretically speaking, all can be put into bgpcfgd, then no j2 template left. But that would be a huge work.